### ISSUE TITLE
Is the binding visible in the `else`?

### ISSUE THREAD

**Issue by bakkot** [tc39/proposal-Declarations-in-Conditionals#3]
Moving from #2.

Consider

```js
if (const a = Math.random() < .5) {
  console.log(a); // definitely works
  console.log(b); // definitely throws
} else if (const b = Math.random() < .5) {
  console.log(a); // works?
  console.log(b); // definitely works
} else {
  console.log(a); // works?
  console.log(b); // works?
}
```

What should the three `works?` lines do?

---

For background, in C++,

```cpp
#include <iostream>
using namespace std;

int main() {
	if (bool a = false) {
		cout << a << endl;
	} else if (bool b = false) {
		cout << a << endl;
		cout << b << endl;
	} else {
		cout << a << endl;
		cout << b << endl;
	}
	return 0;
}
```
is a legal program.

**Comment by dcrousso** [comment]
From the last line of the Overview:
> ... and the variable would only be visible inside the if/while block.

```js
if (const a = Math.random() < 0.5) {
    console.log(a); // works
    console.log(b); // throws
} else if (const b = Math.random() < 0.5) {
    console.log(a); // throws
    console.log(b); // works
} else {
    console.log(a); // throws
    console.log(b); // throws
}
```

As mentioned in [#2](https://github.com/dcrousso/JS-Declarations-in-Conditionals/issues/2#issuecomment-535121495), this is what makes sense to me, but I could be convinced otherwise.  I don't feel particularly strongly either way, other than that it seems weird to have a variable be visible across multiple `{ ... }`.

**Comment by bakkot** [comment]
> other than that it seems weird to have a variable be visible across multiple `{ ... }`

That's the normal behavior of variables: a variable declared in some outer scope is visible to _all_ nested scopes, not just the immediately following one.

**Comment by ljharb** [comment]
I don’t believe almost anybody considers an else or an else if as a nested scope.

**Comment by dcrousso** [comment]
Although this isn't "technically" correct, when I see something like:
```js
if (let data = foo.data) {
    ...
}
```
I think that `data` is actually "part" of the `{ ... }`, not outside it.  If I declared `data` outside of the `if ( ... )`, then yes, it would be visible in both `{ ... }`, but declaring it inside the `if ( ... )` makes my mind see it differently.

I don't consider the `else` (or `else if`) to be nested inside the `if`.  I see it as a completely different scope that's only related because it's easier to write `else` than another `if` with an inverse condition.

I realize that others may see that differently, however, which is why I'm open to alternative ideas, but I'd like a good reason for making it visible inside the `else` other than "that's how C++ does it".

To be clear, this is not meant as a criticism of you or your ideas, but just how I'm approaching this particular idea.

**Comment by hax** [comment]
> I think that data is actually "part" of the `{ ... }`

`if` without `{}` : 
```js
if (let data = foo.data) process(data)
else console.log(`data is falsy: ${data}`)
```

And as [future work](https://github.com/dcrousso/JS-Declarations-in-Conditionals#future-work)
```js
if (let x = 1, y = 2; x || y) {
    /* ... */
}
```
it's likely there will be cases programmers want to access `x` or `y` in `else` clause.

But If we really make `x` visible in `else`, 
```js
let x = ...
if (let x = foo.data) {
  ...
} else if (let x = bar.data) {
  ...
} else {
  console.log(x) // a little bit confusion, I guess some will expect outer x
}
```

Another point is we'd better coordinate with [`try (let x = ...) {} catch {} finally {}`](https://github.com/tc39/proposal-explicit-resource-management) which currently seems do not allow `x` be visible in `catch`/`finally` clauses.

**Comment by waldemarhorwat** [comment]
*Not* extending the scope to the else clause would be highly problematic for a couple reasons:

* Incompability with desirable future extensions such as `if (let x = foo(); bar(x)) { ... } else {console.log(x)}`
* Gratuituous subtle divergence from C++, making it hard to remember which one is which

Note that C++17 [includes both forms](https://en.cppreference.com/w/cpp/language/if) `if (Foo x = ...)` and `if (Foo x = ...; condition)` and extends the scope to both the then and else clauses.

**Comment by hax** [comment]
@waldemarhorwat ~~I think python has issues of `for...else` feature (https://mail.python.org/pipermail/python-ideas/2009-October/006155.html) and I very doubt it could be success in JS. And~~ I don't think C++ consistency should have very high priority.

But if I'm forced to choose one side, i will support extending the scope!

**Comment by michaelficarra** [comment]
There's no way this feature pays for itself if the scope includes both consequent and alternate.

```js
if (let a = b()) {
  // ...
} else {
  // ...
}
```

as sugar for

```js
{
  let a = b();
  if (a) {
    // ...
  } else {
    // ...
  }
}
```

is not a big win for the added language complexity. And limiting the scope to just the consequent brings with it other problems, including the divergence from C++ and confusion with "through" references (references that would have been shadowed if the scope included the alternate). I'd like to stress that I consider myself to be squarely in the target demographic for this proposal, as I am always trying to minimise the scope of my variables, my try/catch statements, etc when writing JavaScript.

**Comment by ljharb** [comment]
Adherence to C++ is irrelevant unless you're a C++ programmer; if there's good reasons to diverge, we should do so.

**Comment by michaelficarra** [comment]
@ljharb Of course, that is good general practice, but we should not ignore that the proposal was presented as essentially "I am a C++ programmer and I found it surprising that JS does not have this feature". Implied in that is that the feature should be familiar for C++ programmers. It's not fatal if it fails that requirement.

**Comment by rkirsling** [comment]
Moreover, alignment with C++ here equates to divergence from [Rust](https://doc.rust-lang.org/book/ch06-03-if-let.html), [Swift](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html#ID333), etc.

I write C++ for my job and make use of this construct, but I'm shocked to learn that it **does** diverge from modern languages on this point. :flushed: (_A coworker pointed out that it could be useful when dealing with C-style return codes; this makes sense for C++, but I don't think this applies much elsewhere..._)

I strongly agree with @michaelficarra that this feature doesn't pay for itself if it's just `if`-in-a-block.

**Comment by dcrousso** [comment]
From @michaelficarra's [comment](https://github.com/dcrousso/JS-Declarations-in-Conditionals/issues/3#issuecomment-537664355):
> @ljharb Of course, that is good general practice, but we should not ignore that the proposal was presented as essentially "I am a C++ programmer and I found it surprising that JS does not have this feature". Implied in that is that the feature should be familiar for C++ programmers. It's not fatal if it fails that requirement.

I apologize if that was how I portrayed it.  It's not my intention to make this identical to C++.  In fact, I didn't realize that C++ makes the variable visible in the `else` until recently, and I personally have yet to find that useful at all.  That is one of the reasons why I initially proposed not doing this for JavaScript (variable is _only_ visible inside `if`).

**Comment by ecyrbe** [comment]
I strongly disagree. C++ construct is the only logical construct for me. I'll vote no with all my power against anything hiding the variable in else clause.
It's an if/else clause block! Not a if only block! Anything declared in the block should be available in the block !
Anyone wanting it otherwise is insane.

**Comment by Mouvedia** [comment]
> What should the three `works?` lines do?

1. throws
2. throws
3. throws

**Comment by ljharb** [comment]
@Mouvedia in strict mode, as opposed to throwing a reference error?

**Comment by Mouvedia** [comment]
<strike>Well I don't know if `a` and `b` were declared before but yeah either output the outer value or throw.</strike>

He's using `const` so a redeclaration would throw anyway. So these are implicitly/probably not declared.

**Comment by ecyrbe** [comment]
> I don’t believe almost anybody considers an else or an else if as a nested scope.

You're wrong. From langage design, the if / else is a single block scope.

People are talking ans confusing a totally different feature. A conditional let if statement like the one in rust, or c# (used explicitly as a simpler pattern matching tool) or the one in Swift (used explicitly to use a conditional variable, witch is a not null pattern matching).
This was not the use case for this proposal.
If you want a pattern matching 'if' construct, i would agree that there would ne no point in having the variable available in the else clause, because in the else clause the pattern dis not match!
If you want a js 'if pattern matching' you should change the syntaxe of this proposal like so:
```javascript
If let { a, b} = c ( a==2) {
  return true;
} else {
 // Pattern matching failed, no reason to have access to a and b
  return false;
}
```
Since we don't yiet have pattern matching in js, the proposal in it's form is a block scoped declaration proposal. And then any declared variable should be available in the else scope.

**Comment by ljharb** [comment]
@ecyrbe not from language usage - and since there’s no way to observe either, it’s up to individual interpretation.

If this proposal is a nerfed form of pattern matching, i would object to it advancing. This proposal, to me, is “declare a binding in the if condition, for use in the if block”. If you want to use a variable in two blocks (inside two adjacent sets of curly braces) you should always need to declare a variable outside both sets of curly braces.

**Comment by bakkot** [comment]
@ljharb The declaration _is_ outside both sets of curly braces. If that's the rule, it should be available in both.

**Comment by ljharb** [comment]
@bakkot the rule is also (arguably) that things declared in attached parens don't make it available outside the immediately adjacent block:
```js
try {
} catch (e) {
} finally {
  // no e available here
}
```

**Comment by ecyrbe** [comment]
@ljharb you are confusing everything. The catch clause is like a pattern matching. e is only valid on that block. The finally block is executed when e can be undefined.
That is not the case for if/else clause.
I can't believe we have to explain these details. Are you really a js teacher ?

**Comment by ljharb** [comment]
Why is that not the case for if/else clauses? The else runs when the if condition is falsy; also, e can be undefined in the catch block.

**Comment by jridgewell** [comment]
> I can't believe we have to explain these details. Are you really a js teacher ?

Personal attacks are not acceptable communication. See https://tc39.es/code-of-conduct/

- - -

> The declaration is outside both sets of curly braces. If that's the rule, it should be available in both.

Yes!

- - - 

> Why is that not the case for if/else clauses? The else runs when the if condition is falsy;

Personally, `catch` is legacy. It's not a `catch (let e) {}`, it's more like a param `catch(e) {}`. I'm willing to let this be a wart with an otherwise consistent feature.

> also, e can be undefined in the catch block.

My `if (const foo = undefined; !foo)` can also have `foo === undefined`, so I don't understand this significance.

**Comment by ecyrbe** [comment]
We have no way to deduce the intent of the declaration in the if/else block. By construct the variable could be used in the else clause. Falsy condition does not mean that the var is invalid in this proposal.
In contrast to catch that gives you the garanty that an exception was throwned. Then e should be valid. If you happen to throw an undefined value it's intentional , like setting a variable to undefined. You juste happen to throw undefined value but in the finally scope you happen to really have an invalid e.

**Comment by puppy0cam** [comment]
If we were to allow all 3 `works?` to work, then one could transpile to something like this:
```JavaScript
{
    const a = Math.random();
    if (a < .5) {
        console.log(a); // definitely works
        console.log(b); // definitely throws
    } else {
        const b = Math.random();
        if (b < .5) {
            console.log(a); // works?
            console.log(b); // definitely works
        } else {
            console.log(a); // works?
            console.log(b); // works?
        }
    }
}
```
However, the untranspiled syntax may not make a lot of sense unless there was a way to declare that you also want to 'share' that variable with an alternative block if it's truthy block is never run.
Such a thing might look like this:
```JavaScript
if (sharing const a = Math.random < .5) {
    console.log(a); // definitely works
    console.log(b); // definitely throws
} else if (sharing const b = Math.random() < .5) {
    console.log(a); // definitely works
    console.log(b); // definitely works
} else if (const c = Math.random() < .5) {
    console.log(a); // works?
    console.log(b); // definitely works
    console.log(c); // definitely works
} else {
    console.log(a); // works?
    console.log(b); // works?
    console.log(c); // definitely throws
}
```
It may also be more correct for a `shared` keyword to be better as a followup proposal rather than packed into this proposal.

**Comment by bramus** [comment]
When declaring variables inside the `initialization` of a `for` loop, variables declared using a LexicalDeclaration (e.g. `let`) are local to the statement.

```js
const cars = ['Ford', 'Mercedes', 'BMW'];

for (let i = 0, len = cars.length; i &lt; len; i++) {
  console.log(len); // ~> 3
  console.log(cars[i]);
}
console.log(len); // ReferenceError: len is not defined
```

To me it'd be logic that Declarations in Conditionals follows the same logic.

**Comment by ghost** [comment]
Well, consider this use:
```js
if ( const a = Math.random(); a < 0.5 ) {
    console.log("Number %d is above 0.5!", a);
} else {
    console.log("Number %d is not above 0.5!", a);
}
```
there, I have a reason for wanting the original value of `a`, but from within the else clause.

In what ways would that be confusing to a JS developer?

Then again, even [Rust doesn't allow visibility to the else clause](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&code=fn%20main()%20%7B%0A%20%20%20%20if%20let%20condition%20%3D%20false%20%7B%0A%20%20%20%20%09println!(%22%5Bif%5D%20condition%20was%20%7B%7D%22%2C%20condition)%3B%0A%20%20%20%20%7D%20else%20%7B%0A%20%20%20%20%09println!(%22%5Belse%5D%20condition%20was%20%7B%7D%22%2C%20condition)%3B%0A%20%20%20%20%7D%0A%7D), so Rust might've been a much better example in the readme, as opposed to C++.

**Comment by ecyrbe** [comment]
> Then again, even [Rust doesn't allow visibility to the else clause](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&code=fn%20main()%20%7B%0A%20%20%20%20if%20let%20condition%20%3D%20false%20%7B%0A%20%20%20%20%09println!(%22%5Bif%5D%20condition%20was%20%7B%7D%22%2C%20condition)%3B%0A%20%20%20%20%7D%20else%20%7B%0A%20%20%20%20%09println!(%22%5Belse%5D%20condition%20was%20%7B%7D%22%2C%20condition)%3B%0A%20%20%20%20%7D%0A%7D), so Rust might've been a much better example in the readme, as opposed to C++.

I'll repeat myself, but RUST  `if/let` statement is a pattern matching statement! not a declaration in conditional.  
It can't be defined in else clause, because pattern matching is a value extraction if matching. In else clause, you did not match, so you have no matched pattern, so no defined variable!   
  
If you  want the same thing, you should wait or help to champion the js **pattern matching proposal** to go further in stages.

**Comment by ghost** [comment]
@ecyrbe Ah, yes, forgive me, I forgot that there was a proposal to add that to JavaScript.
And yes, I agree with Rust's if/let being a form of pattern matching, it checks if the value is in the correct shape, and then uses it as appropriate, otherwise it is unusable.

If the current plan is to deny it to the `if` scope, then this entire proposal is useless.
You might as well use labels
```js
if ( let x = foo(); undefined !== x ) {
    // use `x` however you want
} else {
    // no `x`
}
```
vs
```js
_: {
    let x = foo();
    if ( undefined === x ) {
        break _;
    }

    // use `x` however you want
}
// no `x`
```

Maybe one may not be familiar with the syntax, but it's semanticially equivalent in almost every way; to add a different syntax for the same thing, I cannot support that, unless the older method were to be deprecated and removed.

**Comment by dcrousso** [comment]
ive been giving this a lot of thought lately and it seems like we're perhaps at something of an impasse

im going to try to summarize what i see from the above

some folks say that this is effectively a waste of time if the variable is extended to the `else` as in that case it's basically just syntax sugar for wrapping the variable and `if` inside `{ ... }`

some folks say that this is useless without the variable being extended to the `else` (especially if the more "extended" form with a second expression after a `;` is supported) as it doesnt allow developers to differentiate different kinds of falsy and kind of breaks the "rule" that things outside of `{ ... }` are always available within them

at the very least it does seem like we all agree that being able to declare a variable inside an `if (...)` and being able to use it inside the "pass" block is valuable

i personally lean more towards not exposing in the `else` as it actually provides a new behavior that's not currently possible today without some shenanigans

not to mention, if the variable was exposed in the `else` it would not be possible to do something like this (which btw is very close to some code i wrote super recently)
```js
if (let data = node.firstChild?.data) {
    // ...
} else if (let data = node.lastChild?.data) {
    // ...
} else {
    // ...
}
```

id like to try writing some spec text with the "not in `else`" behavior (as well as a second expression after a `;` to support allowing destructuring since that's just too useful) because i think there's really quite a lot of value in this (and ive never written spec text before so id want to give it a shot)

hopefully ill be done in time to present at the next plenary and we can discuss it more then

**Comment by ljharb** [comment]
That possibility could still be allowed if it was like a virtual scope created by curlies around the construct - which is also what you’d do now to make it available inside the else, so that’s not a new capability.

iow, i think your snippet should be allowed, and also the binding be available in the else. Obviously that adds the difficulty of deciding which `data` (if any) is available in the else.

**Comment by dcrousso** [comment]
oh yeah i mean technically i think it actually ends up being something like this
```js
if (let data = node.firstChild?.data) {
    // ...
} else {
    if (let data = node.lastChild?.data) {
        // ...
    } else {
        // ...
    }
}
```
so it definitely is possible to do it today

as you pointed out, i was more referring to the confusion about which `data` would end up being used if referenced in the final `else`

if we wanted to follow existing similar behavior, it would just be the last `data` like this
```js
let data = 1;
if (true) {
    let data = 2;
    if (true) {
        console.log(data); // 2
    }
}
```
but again i think that's a lot more clear to the developer than if they're all written at the same indentation level (i.e. it's not as obvious that the second `if` is "within" the `else` and therefore would shadows the first `if`)

not exposing it in the `else` avoids this issue entirely

**Comment by bakkot** [comment]
Since this issue was made we've started talking about [extractors](https://github.com/tc39/proposal-extractors).

Extractors can fail to match. In a normal declaration context, that throws. I think there is an obvious way that would work with a declaration in an `if`, which is how it works [in Rust](https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html):

```js
if (let Some(x) = val) {
  // only runs if the extractor succeeded
} else {
  // only runs if the extractor failed 
}
```

Given this behavior, there is no reasonable way to make `x` visible in the `else`, because it doesn't actually have any value if the extractor failed. If we think we might want this - and I think if we're going to do extractors at all then we should also make the above pattern work - then we should not make `x` visible in the `else` in this proposal, even if it goes first.

**Comment by ljharb** [comment]
There's other reasons the extractor would fail besides "having no value" - and everything has a value, even if it's `undefined`.

**Comment by bakkot** [comment]
When an extractor fails, the LHS is never initialized. It literally does not have a value. If you closed over the binding prior to reaching it and then invoked the closure after handling the error, the binding would be in TDZ. With this syntax you can't even do that (unless we make the binding available in the `else`, where it would always be in TDZ, which seems silly).

**Comment by ljharb** [comment]
ah, i see what you mean. I think it'd be perfectly fine for the else visibility to be different when extractors are used, vs when they are not, personally, such that it needn't tie our hands here.

**Comment by bakkot** [comment]
Thinking about #12 also inclines me towards having the binding not visible in the `else`.

Consider the case where you want to acquire a resource and then depending on the state of the resource either do something with it then dispose it, or dispose it immediately and do something else. That is, something like

```js
let neededUpdate;
{
  using resource = getResource();
  if (resource.needsUpdate) {
    neededUpdate = true;
    resource.update();
  }
}
if (!neededUpdate) {
  await postMessage('no updates required');
}
```
This is super awkward, but it's genuinely my best attempt to write this. You can also do it with a DisposableStack:

```js
{
  // we need to use a DisposableStack instead of just calling `resource[Symbol.dispose]()` ourselves to avoid double-disposing 
  using disposer = new DisposableStack();
  const resource = disposer.use(getResource());

  if (resource.needsUpdate) {
    resource.update();
  } else {
    disposer.dispose();
    await postMessage('no updates required');
  }
}
```
but that gets a whole extra data structure involved and isn't really any shorter (but it does let you have the if-else structure that you'd like).

But with this proposal, _if_ the binding is disposed before the `else`, this becomes

```js
if (using resource = getResource(); resource.needsUpdate) {
  resource.update();
} else {
  await postMessage('no updates required');
}
```
which is much, much better.

Of course, if you _want_ the resource available in the `else`, you just put it outside of the `if` (as you already can today):

```js
{
  using resource = getResource();
  if (resource.needsUpdate) {
    resource.update();
  } else {
    await postMessage('no updates required');
  }
}
```
so there's little cost to giving that up.

**Comment by bakkot** [comment]
Rust has something like this feature, and they also have something analogous to `using` (except it's just how all declarations work). In older versions, the value from the condition would live through the `else`, and be disposed after it. They actually [_changed it_](https://doc.rust-lang.org/edition-guide/rust-2024/temporary-if-let-scope.html) in the 2024 edition so that in the `else` case the value is disposed before the `else` is evaluated. They made the change because the old behavior (having the value survive through both arms) was causing [actual bugs in practice](https://brooksblog.bearblog.dev/rusts-sneaky-deadlock-with-if-let-blocks/).

This strikes me as an extremely strong reason to think that at least `using` declarations ([assuming they're allowed](https://github.com/tc39/proposal-Declarations-in-Conditionals/issues/12)) should be disposed before the `else` executes. I think it would be kind of insane for the value from a `using` declaration to be disposed before the `else` but still make the binding available in the `else` - using values after they're disposed is a major footgun, and otherwise only comes up if you leak a closure which captures a `using` binding.

So I think that having the binding not visible in the `else` is the right behavior.

**Comment by bvisness** [comment]
Since it seemingly hasn't been mentioned yet, Go has the declaration be visible in the `else` case: https://go.dev/play/p/Ns0wKw7-zzn

I frequently make use of this with patterns like:

```go
if err := doThing(); err == nil {
  // happy path, no error
} else if err == NotFound {
  // expected error, handle gracefully
} else {
  return fmt.Errorf("unknown error when doing thing: %w", err)
}
```

I can easily imagine many similar cases for JS where it would be useful to have the declaration available in the `else`.

Furthermore, I think it seems very natural for the declaration to be scoped to the entire construct, not just to specific branches of the construct. Also, the Rust example seems motivated by RAII behaviors, specifically the choice of when values get dropped. This does not seem relevant to JS.

**Comment by bakkot** [comment]
I don't see any reason to write that over just putting `let err = doThing()` before the `if`. That seems clearer as well.

> Also, the Rust example seems motivated by RAII behaviors, specifically the choice of when values get dropped. This does not seem relevant to JS.

As mentioned, JS now has `using`, which is basically RAII (or at least has the same destroy-on-scope-exit behavior, which is the relevant similarity).

**Comment by bvisness** [comment]
Fair point, I had forgotten about `using`.

Also, unless I've missed it, there seems to be little discussion so far of whether declarations should be allowed in the `else if`. Some examples in this thread assume that you can, but I'm not sure this has been explicitly addressed. I think this is very relevant to the overall question of scope.

It seems to me that there are only two coherent designs:

1. Only allow declarations in the `if`, not in the `else if`, and these declarations are visible to the entire construct (`if`, `else if`, and `else`).
2. Allow declarations in both `if` and `else if`, and these declarations are only visible to that branch of the construct.

Anything else seems like it would have very confusing scoping behavior.

**Comment by VanillaMaster** [comment]
i think introducing escape hatch for scoping can help
```js
if (const x = foo(); x !== null) {
    // 'x' is visible
} else {
    // 'x' is not defined
}

if (const { x, y } = foo(); x !== null; /* list of names extended to else, just like 'ExportsList' */ x) {
    // 'x' and 'y' are visible
} else {
    // 'x' is visible
    // 'y' is not defined
}

if (const { x, y } = foo(); x !== null; x, y) {
    // 'x' and 'y' are visible
} else {
    // 'x' and 'y' are visible
}

// may be even allow some absurd combinations inherited from export declaration
if (const { x, y } = foo(); x !== null; x, y as z) {
    /* ... */
} else {
    /* ... */
}

if (const { x, y } = foo(); x !== null; *) {
    /* ... */
} else {
    /* ... */
}

```
[ExportsList](https://tc39.es/ecma262/#prod-ExportsList) definition

**Comment by DominoPivot** [comment]
I don't think it makes sense for the presence of an `else` before an `if` to change the allowed syntax of the condition. Take this code for example, with two redundant `else` since each branch is terminating:

```js
if (const err = someTest(value))
    throw Error(`Some error: ${err}`);
else if (const err = anotherTest(value))
    throw Error(`Another error: ${err}`);
else
    return value;
```

If declarations are not allowed in `else if`, the redundant `else` causes a syntax error. And if declarations are allowed in `else if` but all share the same scope, `err` is declared twice which is still a syntax error. If each branch has an independent scope, this example works but plenty of use cases are invalidated. If each branch opens a nested scope, the presence of a redundant `else` before `if` can influence which variable an identifier refers to.

The more I read this thread, the more I think making declarations inside `if` a special case just invites confusion.

**Comment by bakkot** [comment]
`else if` is not a special form in JavaScript; it's just an `else` with an `if` inside of it. As such, I don't think there's really any question about how `else if` works, regardless of whether bindings introduced in the above `if` test are visible: just as `if (let x = ...)` introduces a new scope, in which the introduced binding can shadow outer bindings, an `else if (let x = ...)` would introduce a new scope, in which the introduced binding can shadow outer bindings (including bindings introduced by an earlier `if (let x = ...)`).

Which is to say, your example will be legal regardless of the resolution of this thread.
